[SPARK-11703] [Docker] Fix spark-mesos Dockerfile to use spark 1.5.1 and mesos 0.25

lmtjalves opened the pull request:
Update to downloads spark 1.5.1 and to download it from the closest mirror.

lmtjalves committed:
Update to download spark 1.5.1

Update to downloads spark 1.5.1 and to download it from the closest mirror.

AmplabJenkins commented:
Can one of the admins verify this patch?

lmtjalves commented:
I just found this is wrong. Will fix it

srowen commented:
I see what you're doing and think that's OK, but, honestly I'm still surprised to find these Dockerfiles in Spark. I would prefer Spark not be maintaining this for just this reason. For the same reason Spark EC2 was moved out.

lmtjalves committed:
Susbstitute wget by curl. Some updates.

Substitute wget by curl.
Updated mesos to 0.25 and Ubuntu to 14.04.
Fix error relative to the spark instalation directory.

lmtjalves commented:
Ok, anyway, I fixed some issues in the Dockerfile and updated the ubuntu and the mesos versions.

andrewor14 commented:
@lmtjalves please file a JIRA for this.
@tnachen can you have a look? Is there something we can do to avoid having to manually change this version every release?

tnachen commented:
I don't think there is a good way to keep them in sync, I think part of upgrade mesos version will be requiring to update the docker image here as well.

lmtjalves commented:
@andrewor14 I created an issue on JIRA: SPARK-11703

andrewor14 commented:
@lmtjalves please add `[SPARK-11703]` to the title of this patch and make the title a little more descriptive. See how other patches are opened.

andrewor14 commented:
@tnachen any comments? Why do we change the mesos version to `-ubuntu`?

tnachen commented:
We started to add ubuntu and versions since it wasn't explicit what OS image it was built upon.
@andrewor14 what are these docker images used for? I'm trying to understand if we include them so we can easily launch docker images with spark with the previous released version? Earlier I have a docker image that can build from source so we can build a new docker image with the latest source.

I think the fix is fine but bigger question is that if we want to maintain is we need to put this somewhere in the release notes so whoever is relasing can include this in the release steps.

andrewor14 commented:
It was added to provide docker support for launching Spark on Mesos, remember?: #3074

lmtjalves commented:
@andrewor14 Relative to the #3074, do you know if it works in Mesos 0.25?

dragos commented:
A couple of things while trying to run this on Mesos 0.25:

- `ENV SPARK_HOME /opt/spark` doesn't seem to do anything. It always picks up my local `SPARK_HOME`, which fails on Docker
- I got past that by setting `spark.mesos.executor.home=/opt/spark`. This also failed, because it canonicalizes this path, and it happens that `/opt/spark` is a symlink on my system, and Docker uses the resolved link (which does not exist either).
- I got past this by renaming my symlink, but things are still not working. I'm not sure if this last one is due to my setup, or something is really wrong here. Docker networking might need some tweaks (running docker 1.9).

How did you test this? Any comments on how this should be run (do I need to set `spark.mesos.executor.home`?)

```
15/11/19 16:31:40 INFO Utils: Successfully started service 'driverPropsFetcher' on port 49918.
15/11/19 16:31:40 WARN ReliableDeliverySupervisor: Association with remote system [akka.tcp://sparkDriver@10.0.1.47:62271] has failed, address is now gated for [5000] ms. Reason: [Disassociated] 
Exception in thread "main" java.lang.reflect.UndeclaredThrowableException
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1563)
	at org.apache.spark.deploy.SparkHadoopUtil.runAsSparkUser(SparkHadoopUtil.scala:68)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$.run(CoarseGrainedExecutorBackend.scala:149)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$.main(CoarseGrainedExecutorBackend.scala:250)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend.main(CoarseGrainedExecutorBackend.scala)
Caused by: org.apache.spark.rpc.RpcTimeoutException: Futures timed out after [120 seconds]. This timeout is controlled by spark.rpc.lookupTimeout
	at org.apache.spark.rpc.RpcTimeout.org$apache$spark$rpc$RpcTimeout$$createRpcTimeoutException(RpcEnv.scala:214)
	at org.apache.spark.rpc.RpcTimeout$$anonfun$addMessageIfTimeout$1.applyOrElse(RpcEnv.scala:229)
	at org.apache.spark.rpc.RpcTimeout$$anonfun$addMessageIfTimeout$1.applyOrElse(RpcEnv.scala:225)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:33)
	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcEnv.scala:242)
	at org.apache.spark.rpc.RpcEnv.setupEndpointRefByURI(RpcEnv.scala:98)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$$anonfun$run$1.apply$mcV$sp(CoarseGrainedExecutorBackend.scala:162)
	at org.apache.spark.deploy.SparkHadoopUtil$$anon$1.run(SparkHadoopUtil.scala:69)
	at org.apache.spark.deploy.SparkHadoopUtil$$anon$1.run(SparkHadoopUtil.scala:68)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:415)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1548)
	... 4 more
Caused by: java.util.concurrent.TimeoutException: Futures timed out after [120 seconds]
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
	at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:107)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
	at scala.concurrent.Await$.result(package.scala:107)
	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcEnv.scala:241)
	... 11 more
```

andrewor14 commented:
By the way, maybe this should be changed to 1.6.0 now. Then we need to backport whatever fix we have here into the 1.5 branch.

JoshRosen commented:
If you wanted to update this to 1.6.0, doesn't that need to happen after the 1.6.0 release has been published in order for the artifacts to be in an Apache mirror?

dragos commented:
@lmtjalves do you still plan to work on this PR?

andrewor14 commented:
@JoshRosen but by the time 1.6.0 is released this patch won't be in the release so they'd be using a stale docker file. My view is that this file probably shouldn't even be in the Spark repository so we don't have to maintain it like this. Either that or we just put a placeholder `<SPARK_VERSION>` in the file and require the user to edit it if they want to use it.

JoshRosen commented:
Another factor is the fact that this currently uses a fixed Hadoop version, so users may have have to edit this file anyways if the Hadoop version doesn't match the version of HDFS that they use. Given this, I'm also thinking that a placeholder might be a good approach if we decide to keep this file in Spark.

lmtjalves commented:
Ok I'll add the placeholder. The idea of this PR was mostly to make the current dockerfile work (or at least make more sence).

lmtjalves committed:
Replace the spark version by a placeholder

lmtjalves committed:
Added <HADOOP_VERSION> placeholder

andrewor14 commented:
retest this please

andrewor14 commented on docker/spark-mesos/Dockerfile:
```diff
@@ -17,14 +17,18 @@
 #
 
-FROM mesosphere/mesos:0.20.1
+# Note: This is just a template. You need to replace <SPARK_VERSION> and <HADOOP_VERSION> 
```
@tnachen are there other places in the docs where we should document this? Is there a section for "running Mesos with docker" somewhere?

andrewor14 commented:
@lmtjalves can you address the rest of @dragos' comments? He seemed to have found a few potential problems with the new mesos version.

andrewor14 commented:
Also, maybe it makes sense to rename the docker file `Dockerfile.template` to be more explicit

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47668/
Test FAILed.

lmtjalves committed:
Rename Dockerfile to Dockerfile.template

JoshRosen commented:
@lmtjalves, have you addressed @dragos's comments related to the Mesos version? What's the status of this PR?

JoshRosen commented:
Ping @lmtjalves, any status update here? I think this is really close to being merged. If you don't have time to work on this yourself, that's okay; in that case, just close the PR and someone else can take over if there's interest.

andrewor14 commented:
@dragos @tnachen would one of you be willing to take this up and address the remaining comments?

lmtjalves commented:
@JoshRosen I'm going to close then

dragos commented:
Ok, will do.
